### DISCUSSION TITLE
Is there a way to build hls with particular version of fourmolu?

### DISCUSSION THREAD

**Discussion by jhrcek** [haskell/haskell-language-server#3109] [Q&A] [answered]
Currently I'm installing hls from master to get the latest changes.
I'm installing from source using ghcup command like this.

```bash
ghcup compile hls -g master --ghc 8.10.7 --cabal-update -- -f-floskell -f-ormolu -f-brittany -f-stylishHaskell -f-stan
```
Note that I'm disabling some formatters and other plugins I don't want to use.
Unfortunately I usually end up with different version of formolu compiled into hls (not sure how to tell which one) than the one I'm using on command line.
Is there a way (perhaps cabal flag usable in ghcup command above) to force particular version of fourmolu to be used?

**Comment by fendor** [comment, marked as answer]
How about appending `--constraint "fourmolu==0.8.1.0"`?

**Reply by jhrcek** [reply]
Thanks, I didn't know you can do something like that.
Using `--constraint="fourmolu==0.6.0.0"` got me the closest to what I wanted, but failed due to 

<details>
  <summary>unsolvable constraints (cat .ghcup/logs/cabal.log)</summary>
  <pre>
cabal: Could not resolve dependencies:
[__0] trying: hls-fourmolu-plugin-1.0.3.0 (user goal)
[__1] next goal: fourmolu (dependency of hls-fourmolu-plugin)
[__1] rejecting: fourmolu-0.8.0.0, fourmolu-0.7.0.1, fourmolu-0.7.0.0
(constraint from command line flag requires ==0.6.0.0)
[__1] trying: fourmolu-0.6.0.0
[__2] next goal: Cabal (dependency of fourmolu)
[__2] rejecting: Cabal-3.2.1.0/installed-3.2.1.0 (conflict: fourmolu =>
Cabal>=3.6 && <3.7)
[__2] trying: Cabal-3.6.3.0
[__3] trying: hls-stan-plugin-1.0.0.0 (user goal)
[__4] trying: stan-0.0.1.0 (dependency of hls-stan-plugin)
[__5] next goal: extensions (dependency of stan)
[__5] rejecting: extensions-0.0.0.1 (conflict: Cabal==3.6.3.0, extensions =>
Cabal>=3.0 && <3.3)
[__5] skipping: extensions-0.0.0.0 (has the same characteristics that caused
the previous version to fail: excludes 'Cabal' version 3.6.3.0)
[__5] fail (backjumping, conflict set: Cabal, extensions, stan)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: fourmolu, Cabal, hls-fourmolu-plugin,
stan, extensions, hls-stan-plugin
Try running with --minimize-conflict-set to improve the error message.

  </pre>
</details>

I the conflict is between 
A) hls-stan-plugin -> stan -> extensions package which requires cabal < 3.3
B) hls-fourmolu-plugin -> fourmolu (0.6.0.0) which requires cabal >= 3.6
I wonder why is cabal trying to solve dependencies from stan if I'm explicitly disabling stan plugin via `-f-stan` in the main ghcup command :thinking:

**Reply by jhrcek** [reply]
Wohoo, in the end I was able to make it work by letting cabal solver pick newer version of cabal as dependency of extensions (which is dependency of stan, which I don't want anyway). For reference here's the final command that installs hls with fourmolu version 0.6.0.0 which I wanted:

```bash
ghcup compile hls -g master --ghc 8.10.7 --cabal-update -- -f-floskell -f-ormolu -f-brittany -f-stylishHaskell -f-stan --constraint="fourmolu==0.6.0.0" --allow-newer="extensions:Cabal"
```

**Comment by googleson78** [comment]
The general problem of syncing the tools built into hls with the external ones is additionally also discussed more in this issue - https://github.com/haskell/haskell-language-server/issues/411.

**Comment by georgefst** [comment]
See https://github.com/haskell/haskell-language-server/issues/2827 for an alternative approach.
